### PR TITLE
Allow custom processors to be added as a scoped dependency

### PR DESCRIPTION
**Pull request by chrisza4**
Refer from issue #318 

We can fixed this by turning `SentryMiddleware` to be a [factory-based middleware](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/middleware/extensibility?view=aspnetcore-6.0). But this also required some interfaces changes and now all custom event processors will be injected using `PopulateScope` during `OnEvaluation`.

One concern is that this changes might break behaviour of public method `GetAllEventProcessors` since now it can only be used after scope is evaluated. I'm not sure if that could be a concern.

Alternatively, we can move out from `onEvaluating` hook but I don't know if that is more useful so I just put it there first to make it easier to write test in same pattern (testing `PopulateScope`). Let me know if you have other opinion.

### CODE REVIEW

File: test/Sentry.AspNetCore.Tests/IntegrationsTests.EventProcessor.cs [hunk 1]
```diff
@@ -41,10 +39,9 @@ public async Task SentryEventExceptionProcessor_Transient_LifetimeRespected()
         _ = await HttpClient.GetAsync("/throw");
         _ = await HttpClient.GetAsync("/throw");
 
-        // First resolve only to decided if worth patching SentryOptions
-        Assert.Equal(3, exceptionProcessorsResolved.Count);
-        exceptionProcessorsResolved[0].DidNotReceive().Process(Arg.Any<Exception>(), Arg.Any<SentryEvent>());
+        Assert.Equal(2, exceptionProcessorsResolved.Count);
+        // exceptionProcessorsResolved[0].DidNotReceive().Process(Arg.Any<Exception>(), Arg.Any<SentryEvent>());
```
Comment (chrisza4) [inline test/Sentry.AspNetCore.Tests/IntegrationsTests.EventProcessor.cs:43]: Oops

**Comment by mattjohnsonpint** [comment]
Thanks for this! Looks promising on first glance.  I'll review in more detail next week.

**Comment by SimonCropp** [comment]
wouldn't this result in: processors explicitly added to options being ignored?

**Comment by chrisza4** [comment]
Few test failed. Let me take a look.


@SimonCropp 
> wouldn't this result in: processors explicitly added to options being ignored?

As I read into [Sentry SDK doc](https://docs.sentry.io/platforms/dotnet/guides/aspnetcore/#adding-event-and-exception-processors), I thought we can only use processors via dependency injection. If there is other way then let me know where and I can take a look.

**Comment by mattjohnsonpint** [comment]
During initialization, event processors can be added to the options:

```csharp
.UseSentry(options =>
{
    options.AddEventProcessor(new MyEventProcessor());
}
```

They can also be added for a given scope:

```csharp
SentrySdk.ConfigureScope(scope =>
{
    scope.AddEventProcessor(new MyEventProcessor());
});
```

Both of those would need to continue to work.

**Comment by mattjohnsonpint** [comment]
Looking over the PR, the main issue I see is now *all* processors are registered on the scope.

Instead, let's control this based on how the processor was registered with DI.

- If registered using `AddSingleton`, then it should be applied globally - as if registered directly on the options during initialization.  This is the existing behavior.
- If registered using `AddScoped`, then it should be applied for the current scope - as if registered with `SentrySdk.ConfigureScope`.

**Comment by mattjohnsonpint** [comment]
Actually, ignore my last comments.  There's no need to separate those paths out, as the effect would be the same as the approach you're currently taking.  In other words, registering the same instance of an event processor on each scope is identical to registering it once on options - since we gather both at the same time when we run them.

**Comment by mattjohnsonpint** [comment]
I see we are doing this for event processors and exception processors.  Please also do the same for transaction processors added in #1861.  Looks like we missed this back then.  Thanks.

File: src/Sentry.AspNetCore/SentryMiddleware.cs [hunk 2]
```diff
@@ -162,6 +165,17 @@ private void SyncOptionsScope(IHub newHub)
...
+        {
+            var toAdd = _eventProcessors.Except(scope.GetAllEventProcessors());
+            scope.AddEventProcessors(toAdd);
+        }
+        if (_eventExceptionProcessors != null)
+        {
+            var toAdd = _eventExceptionProcessors.Except(scope.GetAllExceptionProcessors());
+            scope.AddExceptionProcessors(toAdd);
+        }
+
```
Comment (mattjohnsonpint) [inline src/Sentry.AspNetCore/SentryMiddleware.cs:178]: The null checks are not necessary.

```suggestion
        scope.AddEventProcessors(_eventProcessors.Except(scope.GetAllEventProcessors()));
        scope.AddExceptionProcessors(_eventExceptionProcessors.Except(scope.GetAllExceptionProcessors()));
```

**Review by mattjohnsonpint — CHANGES_REQUESTED** [review]

**Comment by mattjohnsonpint** [comment]
@SimonCropp

> wouldn't this result in: processors explicitly added to options being ignored?

No, because `scope.GetAllEventProcessors()` returns both:

https://github.com/getsentry/sentry-dotnet/blob/77ccdb2dc01b85c1dc53604f9186f715f66b9ff8/src/Sentry/ScopeExtensions.cs#L18-L33

**Comment by mattjohnsonpint** [comment]
Needs transaction processors added, and a changelog entry, then LGTM.  Thanks!

**Review by mattjohnsonpint — APPROVED** [review]

**Comment by mattjohnsonpint** [comment]
I added the changelog entry.  We'll do transaction processors in a separate PR.  Thanks for your contribution!

**Pull request merged**